src/libraries/System.Text.Json/docs/ThreatModel.md

```diff
@@ -0,0 +1,206 @@
+
+The message for an exception thrown by the serializer may contain the value or partial value of the current JSON property (on deserialization) or CLR property (on serialization) being processed. This is direct “user data” exposed; without this information it would be more difficult to determine what the errors are. It is expected that any application that displays or saves this information treats it as raw text.
+
+#### Mitigation
+
+- When using the serializer, consider what information might be included if an exception is thrown when handling your input data.
+
+- Be cautious about how the raw data from the exception message is processed or forwarded by your application.
+
+### Threat: Errors due to badly constructed, malicious, or non-RFC-compliant JSON payloads
+
+Malicious payloads may contain badly formatted JSON in an attempt to cause unexpected exceptions, put the serializer or reader in an invalid state, or cause other security-sensitive problems.
+
+The `JsonWriterOptions.SkipValidation` property specifies whether to allow a caller to write invalid JSON using `Utf8JsonWriter`. For better performance when writing, `JsonSerializer` sets this property to `true`. This setting is used even when a user-provided converter is used to serialize a type. This means that the only safeguard the `JsonSerializer.Serialize` offers against writing invalid JSON is ensuring that the nesting level of the written JSON is the same when the converter is finished as when it started. This limited validation opens the possiblility of writing invalid JSON that may be accidently (or intentionally) misused and presents a threat when passed as an input to parsers such as `JsonSerializer.Deserialize`.
+
+By default, the `JsonSerializer` and `Utf8JsonReader` strictly honor the [RFC 8259 specification](https://tools.ietf.org/html/rfc8259). When enabling non-RFC-compliant behaviors (such as permitting single or multi-line comments when reading JSON), different deserializers may interpret incoming payloads differently. This could result in two different deserializers reading the exact same payload but populating an object's members with different values. This may have security implications for distributed systems. For example, if a validating frontend uses one deserializer but the backend system uses a different deserializer, this may present an opportunity for an adversary to craft a payload which appears safe when seen by the frontend but which is malicious when deserialized by the backend.
+
+#### Mitigation
+
+- The `Utf8JsonReader`, `Utf8JsonWriter`, and `JsonSerializer` types adhere to the [RFC 8259 specification](https://tools.ietf.org/html/rfc8259) by default, requiring the input and output payloads follow the standard formatting for strings, numbers, lists, and objects. Payloads that don't conform will cause a clear `JsonException` to be thrown when reading.
+
+- Avoid writing custom converters incorrectly, particularly for serialization.
```

Comment:
Serialization is a bit weird here imho
The threat description is about malicious payloads, meaning deserialization. Then it starts with serialization and that it is limited to nesting and rfc compliance etc and describes that it is a threat to ``Deserialize``
The mitigation then describes ... particularly for **serialization**. I would have expected that **deserialization** is the offender here where my extra smart code is doing not so smart code (collisions, etc., similar to the stuff that was listed before).  If serialization is the big threat to my own code written with custom converters, then the description above does not make it clear to me.

The second mitigation step about understanding how processing is fine.

Reply:
Well it's probably both, serialization here, to my mind, covers both. So both could be mentioned, rather than just changing it to deserialization.

Reply:
The way I interpreted this was "avoid writing serialization converters that produce non-standard output and thus require the receiver to opt in to potentially dangerous behaviors." But perhaps the wording can be cleaned up a bit to better describe how both serializers and deserializers need to play well w.r.t. standards-compliance?

Reply:
> "avoid writing serialization converters that produce non-standard output and thus require the receiver to opt in to potentially dangerous behaviors."

Yup this is what I was trying to say. I've updated the bullet, and added another one for guidance on deserialization converters.